akka-actor-typed-tests/src/test/scala/akka/actor/typed/testkit/EventFilter.scala

```diff
@@ -0,0 +1,474 @@
+package akka.actor.typed.testkit
+
+/*
+ * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+import java.util.concurrent.LinkedBlockingQueue
+
+import akka.actor.ActorSystem
+import akka.event.Logging
+import akka.testkit.{ TestKit, TestKitExtension }
+import akka.util.BoxedType
+import org.slf4j.event.{ Level, LoggingEvent, SubstituteLoggingEvent }
+
+import scala.concurrent.duration.Duration
+import scala.reflect.ClassTag
+import scala.util.matching.Regex
+
+/**
+ * Implementation helpers of the EventFilter facilities: send `Mute`
+ * to the TestEventListener to install a filter, and `UnMute` to
+ * uninstall it.
+ *
+ * You should always prefer the filter methods in the package object
+ * (see `akka.testkit` `filterEvents` and `filterException`) or on the
+ * EventFilter implementations.
+ */
+/**
+ * Facilities for selectively filtering out expected events from logging so
+ * that you can keep your test run’s console output clean and do not miss real
+ * error messages.
+ *
+ * See the companion object for convenient factory methods.
+ *
+ * If the `occurrences` is set to Int.MaxValue, no tracking is done.
+ */
+abstract class EventFilter(occurrences: Int) {
```

Comment:
How far has this been tried? I see that it is used in a few tests with a SubstituteLogger. What is the idea for testing logging to any Logger (e.g. one given by ActorContext.log)?

Have we looked at any of existing testing utilities? Could it be an alternative to use something external instead of implementing this ourselves?
https://stackoverflow.com/questions/29076981/how-to-intercept-slf4j-logging-via-a-junit-test

Reply:
That's the approach I took. @raboof point me to that very same link, and it was quite useful; I have to say.

Reply:
Ok, I just wanted to make sure that we invented the wheel again. I thought that we maybe could use http://projects.lidalia.org.uk/slf4j-test/ but there seems to be a few problems with that library:
* not maintained (last update 4 years ago)
* no license
* primarily intended for thread local logging (but also supports global log event recording)

Seems ok to continue on the approach taken here with the SubstituteLogger and our own testing api to assert the events. When we have confirmed that it works we should spend some time on the actual api of EventFilter, such as:
* javadsl/scaladsl
* simplify, remove unnecessary cruft

Reply:
:) yes. I tried to resist to implement ourselves but didn't found anything suitable.

Glad you agree on the approach. Let's do as you suggest. I'm looking forward to it.